import-automation/executor/app/service/github_api.py

```diff
@@ -0,0 +1,374 @@
+
+    def _path_containing_file(self, commit_sha: str, dir_path: str,
+                              containing: str) -> bool:
+        """Checks if the directory in the repository contains a file.
+
+        Example:
+            There is a README.md in foo/bar when commit '21d12fr2' happens.
+            The method call
+                _path_containing_file('21d12fr2', 'foo/bar', 'README.md')
+            returns True.
+
+        Args:
+            commit_sha: Commit ID as a string.
+            dir_path: Path of the directory to search in as a string.
+            containing: Name of the file to search for as a string.
+
+        Returns:
+            True if the directory contains a file with the given name. False,
+            otherwise.
+        """
+        files_in_dir = self._query_files_in_dir(commit_sha, dir_path)
+        for filename in files_in_dir:
+            if filename == containing:
+                return True
+        return False
+
+    def _find_dirs_up_path_containing_file(self, commit_sha: str, dir_path: str,
+                                           containing: str,
+                                           searched: Set[str]) -> Set[str]:
+        """Given a path pointing at a directory in the repository, searches for
+        a file in the directory and its parent directories.
+
+        Example:
+            Assume the repository has a subdirectory called 'scripts' and
+            'scripts' has a subdirectory called 'us_fed'.
+            There is a 'README.md' in the root directory of the repository and
+            another in 'us_fed'.
+
+            The method call
+                _find_dirs_up_path_containing_file(
+                    'dw1f231asdf1321', scripts/us_fed', 'README.md', set())
+            first searches in 'us_fed' and adds 'scripts/us_fed' to
+            'found_dirs'. It then searches in 'scripts' and the root directory
+            of the repository.
+            {'scripts/us_fed', ''} is returned.
+
+        Args:
+            commit_sha: Commit ID as a string.
+            dir_path: Path to the directory where the search begins as a string.
+            containing: Filename to search for as a string.
+            searched: Set object containng paths of directories to skip over.
+                If a directory is already in 'searched', it and its parent
+                directories will not be searched. Directories searched this
+                time are also added to it.
+
+        Returns:
+            A set of directory paths each as a string.
+        """
+        found_dirs = set()
+        while dir_path:
+            if dir_path in searched:
+                return found_dirs
+            searched.add(dir_path)
+            if self._path_containing_file(commit_sha, dir_path, containing):
+                found_dirs.add(dir_path)
+            dir_path = os.path.dirname(dir_path)
+
+        # Search root
+        if '' not in searched:
```

Comment:
Does modifying the while condition above to `while dir_path or dir_path == ''` 

simplify the code?  `os.path.dirname('') == ''` so I guess it would loop around again, finding '' in searched and stop.  Maybe this is too tricky.

Reply:
The tests pass. changed